### PR TITLE
API: promote sys/util.h to experimental API

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
As discussed in https://github.com/zephyrproject-rtos/zephyr/pull/24556#issuecomment-624776302 and previously in that PR, we're relying on the definitions in this header as core Zephyr functionality.

It therefore makes sense to make this header documented API so we can start to lock down its stability level in this and subsequent releases.

### CODE REVIEW

File: doc/releases/release-notes-2.4.rst [hunk 1]
```diff
@@ -37,6 +37,10 @@ API Changes
 * Moved to using C99 integer types and deprecate Zephyr integer types.  The
   Zephyr types can be enabled by Kconfig DEPRECATED_ZEPHYR_INT_TYPES option.
 
+* The ``<sys/util.h>`` header has been promoted to a documented API with
+  :ref:`experimental stability <api_lifecycle>`. See :ref:`util_api` for an API
```
Comment (mbolivar-nordic) [inline doc/releases/release-notes-2.4.rst:41]: I picked this as the most conservative choice, but perhaps "Unstable" would be better. LMK!

File: include/sys/util.h [hunk 2]
```diff
@@ -262,106 +341,109 @@ uint8_t u8_to_dec(char *buf, uint8_t buflen, uint8_t value);
...
+ *     #if defined(CONFIG_FLAG) && (CONFIG_FLAG == 1)
+ *     #define MACRO(x) uint32_t x;
+ *     #else
+ *     #define MACRO(x)
+ *     #endif
+ *     MACRO(x)
+ *
+ * However, the advantage of COND_CODE_1() is that code is resolved in
+ * place where it is used, while the \#if method defines MACRO on line,
+ * and requiring it to be invoked again on a separate line. This makes
```
Comment (pabigot) [inline include/sys/util.h:376]: ~defines MACRO on line and requiring it to be invoked again~

defines MACRO on one line and requires it to be defined again

?

**Review by pabigot — CHANGES_REQUESTED** [review]
I agree this should be documented, and is definitely not "stable", so thanks for doing this.  (And for moving the more egregious abuse of the preprocessor to `util_internal.h` so even "advanced" Zephyr users shouldn't feel ashamed for not understanding `<sys/util.h>`....)

One documentation concern that should be addressed.

As a note: A large number of macros in this file appear to derive from https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms#deferred-expression though I don't know whether that's original work (it is the source of https://stackoverflow.com/a/12540675 which is referenced in the header).

~At least some those macros (e.g. `UTIL_IIF`)~ Not sure all macros are used in Zephyr.  Some cleanup and/or provenance documentation may be appropriate.  Not in this PR.

File: include/sys/util.h [hunk 3]
```diff
@@ -28,98 +28,168 @@
...
+ * @brief The smaller value between "a" and "b"
+ * @note Arguments are evaluated twice.
  */
 #ifndef MIN
+/* Use Z_MIN for a GCC-only, single evaluation version */
 #define MIN(a, b) (((a) < (b)) ? (a) : (b))
 #endif
 
+/**
+ * @brief is "x" a power of two?
```
Comment (nordic-krch) [inline include/sys/util.h:177]: `Is "x" ...` ?

File: include/sys/util.h [hunk 4]
```diff
@@ -260,106 +341,109 @@ uint8_t u8_to_dec(char *buf, uint8_t buflen, uint8_t value);
...
+ * However, the advantage of COND_CODE_1() is that code is resolved in
+ * place where it is used, while the \#if method defines MACRO on line,
+ * and requiring it to be invoked again on a separate line. This makes
+ * COND_CODE_1() more useful within recursive macro expansion.
  *
+ * @note _flag can also be a result of preprocessor output, e.g.
+ *	 the expansion of NUM_VA_ARGS_LESS_1(...).
+ *
+ * @param _flag		Evaluated flag
+ * @param _if_1_code	Expanded result if _flag is defined and expands to
```
Comment (nordic-krch) [inline include/sys/util.h:383]: maybe it is worth adding information "must be in parentheses" for each argument in those macros. It is mentioned only once in COND_CODE_1 and never repeated in other macros.

**Review by nordic-krch — COMMENTED** [review]
thanks for doing this!

**Comment by mbolivar-nordic** [comment]
> As a note: A large number of macros in this file appear to derive from https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms#deferred-expression though I don't know whether that's original work (it is the source of https://stackoverflow.com/a/12540675 which is referenced in the header).

IANAL but this post was edited in 2017, which is after [this policy](https://stackoverflow.com/questions/12447557/can-we-have-recursive-macros/12540675#12540675) went into effect, so we appear to be compliant because of [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/26134/files#diff-418e933b5df2b11420aa260d71e1fd5dR91), though a license expert should confirm if that continues to be a concern. Of course, this PR isn't adding that code, just moving it around.

**Comment by mbolivar-nordic** [comment]
@pabigot @nordic-krch thanks for reviews. I've addressed your comments and added a few more cleanups of my own. Notably, I'm adding a removal of the deprecated MACRO_MAP() macro.

**Comment by pabigot** [comment]
> IANAL but this post was edited in 2017, which is after [this policy](https://stackoverflow.com/questions/12447557/can-we-have-recursive-macros/12540675#12540675) went into effect, so we appear to be compliant because of [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/26134/files#diff-418e933b5df2b11420aa260d71e1fd5dR91), though a license expert should confirm if that continues to be a concern. Of course, this PR isn't adding that code, just moving it around.

The links above don't make sense to me: the first is to the same stackoverflow answer that's referenced in the header, not a (Zephyr?) policy; and the second to the entire set of changes in this PR, not a comment.

Regardless, I'm not saying there's something to be done in this PR, just that the provenance of the code in that header and its licensing is not entirely clear.  The copyright assignment to Wind River Systems 2011-2014 as indicated in current master is pretty obviously wrong, though.  At least some of this was originally added in 0a4808b75ca2b8a8836eccd4d5a2069abebd6181 via #4692.  https://stackoverflow.com/help/licensing is also relevant to the question (but not this PR); the `SPDX-License-Identifier` perhaps should be `(Apache-2.0 AND CC-BY-SA-3.0)`.

I'll just make this comment an issue in its own right so we can think about it while we carry on with more immediate concerns.

**Comment by mbolivar-nordic** [comment]
> The links above don't make sense to me: the first is to the same stackoverflow answer that's referenced in the header, not a (Zephyr?) policy; and the second to the entire set of changes in this PR, not a comment.

Sorry, wrong link: https://meta.stackexchange.com/questions/271080/the-mit-license-clarity-on-using-code-on-stack-overflow-and-stack-exchange

In particular:

> Starting February 1, 2016, contributions across the network will be licensed to the public under the following terms:
>
>    - Non-code contributions will continue to be available for use under the terms of CC-BY-SA
>    - Code contributions will be available for use under the terms of the MIT License
>    - You don’t have to include the full MIT License in your code base. Contributors agree to give code users permission to ignore the MIT License’s notice preservation requirement, as long as users give reasonable attribution upon request of the copyright holder (or Stack Exchange on behalf of the contributor). This optional exception to the MIT License will live in our terms of service.
>
> That last bullet makes using Stack Exchange easy, and provides added protection to code contributors and users.
>
> For the vast majority of developers, the above is Too Much Information. The new licensing terms don’t change anything. You can do everything you did before and maybe more when you visit Stack Overflow: copy code, tinker, fiddle, put it in your project, and keep building.
>
>But to future-proof your work, we recommend you do one of these 2 things, or both:
>
>   A) Add a comment to your code that links back to the post where you found it, or
>   B) Comply with the MIT as it’s typically used, by including the full license text in your source

**Review by pabigot — APPROVED** [review]
I'd like to see `kHz` but about 30% of existing usage in Zephyr gets it wrong so it's not a big deal.

**Review by nordic-krch — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
I like the changes.

File: include/sys/util.h [hunk 5]
```diff
@@ -30,98 +35,166 @@
...
 static inline bool is_power_of_two(unsigned int x)
 {
 	return (x != 0U) && ((x & (x - 1)) == 0U);
 }
 
+/**
+ * @brief Arithmetic shift right
+ * @param value value to shift
+ * @param shift number of bits to shift
+ * @return "value" shifted right by "shift"; opened bit positions are
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply include/sys/util.h]: > Not sure it's in scope but `"value"` could be replaced by `@p value` unless some post-Doxygen processing breaks with that usage.

I replaced a bunch of stuff with `@p` throughout the file.

**Comment by mbolivar-nordic** [comment]
DNM today (June 16) until @nordic-krch and @de-nordic have a chance to take another look, since I changed a bunch of language while adding all the `@p` and `<tt>...</tt>`.

**Review by nordic-krch — APPROVED** [review]
looks great

**Review by jfischer-no — APPROVED** [review]

**Comment by nordic-krch** [comment]
@carlescufi do you think we could merge it?

**Review by de-nordic — APPROVED** [review]
Looks OK. Thanks.

**Review by gmarull — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Looks great to me in its current state.

**Pull request merged**